src/cargo/core/resolver/context.rs

```diff
@@ -160,23 +161,32 @@ impl Context {
             }
         }
         debug!("checking if {} is already activated", summary.package_id());
-        if opts.features.all_features {
-            return Ok(false);
-        }
-
-        let has_default_feature = summary.features().contains_key("default");
-        Ok(match self.resolve_features.get(&id) {
-            Some(prev) => {
-                opts.features.features.is_subset(prev)
-                    && (!opts.features.uses_default_features
-                        || prev.contains("default")
-                        || !has_default_feature)
-            }
-            None => {
-                opts.features.features.is_empty()
-                    && (!opts.features.uses_default_features || !has_default_feature)
+        match &opts.features {
+            // This returns `false` for CliFeatures just for simplicity. It
+            // would take a bit of work to compare since they are not in the
+            // same format as DepFeatures (and that may be expensive
+            // performance-wise). Also, it should only occur once for a root
+            // package. The only drawback is that it may re-activate a root
+            // package again, which should only affect performance, but that
+            // should be rare. Cycles should still be detected since those
+            // will have `DepFeatures` edges.
+            RequestedFeatures::CliFeatures(_) => return Ok(false),
```

Comment:
I don't know if this is actually correct.  I didn't fully understand why the original code returned `false` when `all_features` was set.  

This function was originally added in https://github.com/rust-lang/cargo/commit/06f37a09035e8b8a1b9e9f3b538fa1f0d2c63fdf to deal with cycles. It treated `Method::Everything` as "false".  The check for `all_features` as added in https://github.com/rust-lang/cargo/commit/7de30dd21a6b78c718c9ee88b661bd0575542884 without an explanation.

Reply:
I'm unfortunately unable to piece together more context than what you probably already know. I don't really know what this check was doing prior other than exactly what you're doing here which is to generate a little bit more work for the resolver to avoid trying to handle more at this call-site.

I think that's fine, however, and I doubt this will have any noticeable impact on perf or anything like that.